### PR TITLE
Mtl updates

### PR DESCRIPTION
**Pull request by bhancock8**
Address comments from 7/25 feedback session (mostly additional explanatory text).

### CODE REVIEW

File: mtl/multitask_tutorial.py [hunk 1]
```diff
@@ -3,12 +3,16 @@
 # # Multi-Task Learning (MTL) Basics Tutorial
 
 # %% [markdown]
-# Multi-task learning is becoming a standard tool for the modern ML practioner. A major requirement of this approach is the ability to easily *add new datasets, label sets, tasks, and metrics* (and just as easily remove them). Thus, in the Snorkel multi-task model design, each of these concepts have been decoupled.
+# Multi-task learning is becoming a standard tool for the modern ML practioner.
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: What happened to the notebook?

File: mtl/multitask_tutorial.py [hunk 2]
```diff
@@ -3,12 +3,16 @@
...
+# (See Sebastian Ruder's [survey](http://ruder.io/multi-task/) from 2017 for a nice overview).
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: nit: awkward as separate parenthetical sentence

File: mtl/multitask_tutorial.py [hunk 3]
```diff
@@ -3,12 +3,16 @@
...
+# It often leads to computational gains (one model performing many tasks takes up less memory and storage) as well as performance gains (learning to do well on a related _auxiliary_ task can improve the model's ability on the _primary_ task).
+# A major requirement of multi-task learning in general is the ability to easily *add new datasets, label sets, tasks, and metrics* (and just as easily remove them). Thus, in the Snorkel multi-task model design, each of these concepts have been decoupled.
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: > in the Snorkel multi-task model design

the what? shouldn't we say Snorkel comes with a pytorch-based modeling framework that can do MTL first? and position it relative to other things? (not necessary to use but...)

File: mtl/multitask_tutorial.py [hunk 4]
```diff
@@ -3,12 +3,16 @@
...
 
 # %% [markdown]
-# The purpose of this tutorial is to introduce the basic interfaces and flow of the multi-task learning tools within Snorkel (we assume that you have prior experience with MTL, so we don't motivate or explain multi-task learning at large here.
+# The purpose of this tutorial is to introduce the basic interfaces and flow of the multi-task learning tools within Snorkel (we assume that you have prior experience with MTL, so we don't motivate or explain multi-task learning at large here).
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: parens -> separate sentence...it's important

File: mtl/multitask_tutorial.py [hunk 5]
```diff
@@ -29,81 +33,64 @@
 # Our data points are 2D points in a square centered on the origin.
 # Our tasks will be classifying whether these points are:
 #
-# 1. Inside a **unit circle** centered on the origin
-# 2. Inside a **unit square** centered on the origin
+# 1. Inside a **unit circle** centered on the origin (0 = False, 1 = True)
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: label 0...

File: mtl/multitask_tutorial.py [hunk 6]
```diff
@@ -29,81 +33,64 @@
...
+# 2. Inside a **unit square** centered on the origin (0 = False, 1 = True)
 #
 # We'll visualize these decision boundaries in a few cells.
 #
-# _Note: In this toy example, we don't expect these specific tasks to help each other learn, but this is often a benefit of joint training in MTL settings._
+# _Note: In this toy example, we don't expect these specific tasks to help the model improve on the other task, but this is often a benefit of joint training in MTL settings._
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: More confusing than before

File: mtl/multitask_tutorial.py [hunk 7]
```diff
@@ -29,81 +33,64 @@
...
 R = 1  # Unit distance
 
-# Dataset 0 ("circle")
-circle_data = np.random.uniform(0, 1, size=(N, 2)) * 2 - 1
-circle_labels = circle_data[:, 0] ** 2 + circle_data[:, 1] ** 2 < R
-
-# Dataset 1 ("square")
-square_data = np.random.uniform(0, 1, size=(N, 2)) * 2 - 1
-square_labels = (abs(square_data[:, 0]) < 0.5) * (abs(square_data[:, 1]) < 0.5)
+circle_data_splits, circle_label_splits = make_circle_data(N, R)
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: too opaque for this. can we just do `(circle_train, circle_test), (circle....`

File: mtl/multitask_tutorial.py [hunk 8]
```diff
@@ -29,81 +33,64 @@
...
-# %% [markdown]
-# Note that, as is the case throughout the Snorkel repo, the **label -1 is reserved for abstaining/no label**; all actual labels have non-negative integer values: 0, 1, 2, .... This provides flexibility for supervision sources to label only portions of a dataset, for example.
+# %%
+print(f"Data splits: {list(circle_data_splits.keys())}")
+print(f"Training data shape: {circle_data_splits['train'].shape}")
+print(f"Label space: {set(circle_label_splits['train'])}")
 
 # %% [markdown]
 # And we can view the ground truth labels of our tasks visually to confirm our intuition on what the decision boundaries look like.
+# These two datasets occupy the same region in Euclidean space, but are disjoint sets of data points.
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: delete

File: mtl/multitask_tutorial.py [hunk 9]
```diff
@@ -29,81 +33,64 @@
...
+# In the plots below, the purple points represent class 0 and the yellow points represent class 1.
 
 # %%
 import matplotlib.pyplot as plt
 
 fig, axs = plt.subplots(1, 2)
 
-axs[0].scatter(circle_data[:, 0], circle_data[:, 1], c=circle_labels)
+scatter = axs[0].scatter(
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: This code is getting a bit too fancy and hard to read for a simple example. try to be more explicit and work with simple objects (e.g. ndarrays, not dictionaries of them)

File: mtl/multitask_tutorial.py [hunk 10]
```diff
@@ -29,81 +33,64 @@
...
-# We'll now divide the synthetic `circle_data` and `square_data` into train/valid/test splits using our helper function `split_data()`.
-
-# %%
-circle_data_splits, circle_label_splits = split_data(
-    circle_data, circle_labels, seed=SEED
-)
-square_data_splits, square_label_splits = split_data(
-    square_data, square_labels, seed=SEED
-)
+# Note that, as is the case throughout the Snorkel repo, all ground truth labels have non-negative integer values: 0, 1, 2, etc. **The label -1 is reserved for abstaining/no label.** This ability to abstain provides flexibility for supervision sources to label only portions of a dataset (abstaining on the rest), for example.
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: Probably not necessary here

File: mtl/multitask_tutorial.py [hunk 11]
```diff
@@ -141,17 +128,17 @@ def split_data(data, labels, seed=123):
 # ## Define Model
 
 # %% [markdown]
-# Now we'll define the `SnorkelClassifier`, which is build from a list of `Tasks`.
+# Now we'll define the `SnorkelClassifier`, a flexible multi-task classifier written in PyTorch that is built from a list of `Tasks`.
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: we'll instantiate it?

File: mtl/multitask_tutorial.py [hunk 12]
```diff
@@ -141,17 +128,17 @@ def split_data(data, labels, seed=123):
...
 #
-# To specify this sequence of modules, each `Task` defines a **module pool** (a set of modules that it relies on) and a **task flow**—a sequence of `Operation`s. Each `Operation` defines a module and the inputs to feed to that module. These inputs are described with a list of tuples, where each tuple is either (`_input_`, \[field_name\]), or the name of a previous operation and the index of its output to use (most modules have only a single output, so the second element of these tuples is almost always 0).
+# To specify this sequence of modules, each `Task` defines a **module pool** (a set of modules that it relies on) and a **task flow**—a sequence of `Operation`s. Each `Operation` defines a module and the inputs to feed to that module. These inputs are described with a list of tuples, where each tuple has the name of a previous operation (or the keyword `_input_` to denote the original input) and either the name of a field (e.g., if the output of that operation is a `dict`) or an index (e.g., if the output of that operation is a `list` or `tuple`). Most PyTorch modules output a single tensor, so most of the time, the second element of this tuple is 0.
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: Let's point to the API reference here. Or maybe a diagram. I can't make it through this block of text

File: mtl/multitask_tutorial.py [hunk 13]
```diff
@@ -313,8 +335,12 @@ def split_data(data, labels, seed=123):
 # Using the `square_task` definition as a template, fill in the arguments for an `inverse_circle_task` that consists of the same `base_mlp` module as the other tasks and a separate linear head with an output of size 2.
 
 # %%
-# Uncomment and fill in the arguments to the Task object to define the inverse_circle task.
-# inv_circle_task = Task()
+# Uncomment and fill in the arguments to creat a Task object for the inverse_circle task.
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: creat

File: mtl/multitask_tutorial.py [hunk 14]
```diff
@@ -295,10 +317,10 @@ def split_data(data, labels, seed=123):
 # - The Y_dict should map a task name to a set of labels. This will tell the model what path through the network to use when making predictions or calculating loss on batches from this dataset. At this point we haven't yet defined our
 
 # %%
-X_dict = {}  # Replace this with the correct definition
-Y_dict = {}  # Replace this with the correct definition
+X_dict = {}  # Update this line
```
Comment (henryre) [inline mtl/multitask_tutorial.py]: "Update this line" sounds like an internal TODO

Comment (henryre) [reply mtl/multitask_tutorial.py]: "Fill in"?

File: mtl/multitask_tutorial.py (hunk 12 above)
Comment (bhancock8) [reply mtl/multitask_tutorial.py]: It's a little clearer now. Don't think the API will help; it largely repeats this, but in less space. A diagram is probably what we need. Suggesting we merge this now though to unblock the PRs stacked on top of it. Adding the diagram to our wishlist.

**Comment by bhancock8** [comment]
All comments addressed, ready for another look. The commit history is gnarly since the PR was first posted a couple weeks ago, but the last commit is a merge with master and the commits will all be squashed anyway.

File: multitask/utils.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import numpy as np
+from sklearn.model_selection import train_test_split
+
+
+def make_circle_dataset(N, R, **kwargs):
```
Comment (henryre) [inline multitask/utils.py]: why caps?

Comment (henryre) [reply multitask/utils.py]: also typing plz

File: multitask/utils.py [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    return split_data(X, Y, **kwargs)
+
+
+def make_square_dataset(N, R, **kwargs):
+    X = np.random.uniform(0, 1, size=(N, 2)) * 2 - 1
+    Y = ((abs(X[:, 0]) < R / 2) * (abs(X[:, 1]) < R / 2)).astype(int)
+    return split_data(X, Y, **kwargs)
+
+
+def split_data(X, Y, splits=[0.8, 0.1, 0.1], seed=123):
```
Comment (henryre) [inline multitask/utils.py]: no mutable defaults

File: multitask/multitask_tutorial.py [hunk 17]
```diff
@@ -3,24 +3,47 @@
 # # Multi-Task Learning (MTL) Basics Tutorial
 
 # %% [markdown]
-# Multi-task learning is becoming a standard tool for the modern ML practioner. A major requirement of this approach is the ability to easily *add new datasets, label sets, tasks, and metrics* (and just as easily remove them). Thus, in the Snorkel multi-task model design, each of these concepts have been decoupled.
+# Multi-task learning, or training a single model on multiple tasks, is becoming a standard tool for the modern ML practioner (see Ruder's [survey](http://ruder.io/multi-task/) from 2017 for a nice overview).
```
Comment (henryre) [inline multitask/multitask_tutorial.py]: first name

Comment (henryre) [reply multitask/multitask_tutorial.py]: practioner

File: multitask/multitask_tutorial.py [hunk 18]
```diff
@@ -3,24 +3,47 @@
...
+# It often leads to computational gains (one model performing many tasks takes up less memory and storage) as well as performance gains (learning to do well on a related _auxiliary_ task can improve the model's ability on the _primary_ task).
+#
+# While the primary purpose of the Snorkel project is to support training data creation and management, it also comes with a PyTorch-based modeling framework intended to support flexible multi-task learning (e.g. slice-aware models).
+# Using this particular framework (as opposed to other excellent third party libraries) is entirely optional, but we have found it helpful in our own work and so provide it here.
```
Comment (henryre) [inline multitask/multitask_tutorial.py]: libs like....

File: multitask/multitask_tutorial.py [hunk 19]
```diff
@@ -3,24 +3,47 @@
...
+# In particular, because MTL in general often requires easily *adding new datasets, tasks, and metrics* (and just as easily removing them), each of these concepts have been decoupled in the snorkel MTL classifier.
```
Comment (henryre) [inline multitask/multitask_tutorial.py]: Snorkel

File: multitask/multitask_tutorial.py [hunk 20]
```diff
@@ -29,82 +52,59 @@
...
 
-N = 500  # Data points per dataset
+N = 1000  # Data points per dataset
 R = 1  # Unit distance
 
-# Dataset 0 ("circle")
-circle_data = np.random.uniform(0, 1, size=(N, 2)) * 2 - 1
-circle_labels = circle_data[:, 0] ** 2 + circle_data[:, 1] ** 2 < R
+circle_train, circle_valid, circle_test = make_circle_dataset(N, R)
+(circle_X_train, circle_Y_train) = circle_train
```
Comment (henryre) [inline multitask/multitask_tutorial.py]: nit: don't need parens

File: multitask/multitask_tutorial.py [hunk 21]
```diff
@@ -29,82 +52,59 @@
...
+(circle_X_valid, circle_Y_valid) = circle_valid
+(circle_X_test, circle_Y_test) = circle_test
 
-# Dataset 1 ("square")
-square_data = np.random.uniform(0, 1, size=(N, 2)) * 2 - 1
-square_labels = (abs(square_data[:, 0]) < 0.5) * (abs(square_data[:, 1]) < 0.5)
+square_train, square_valid, square_test = make_square_dataset(N, R)
+(square_X_train, square_Y_train) = square_train
```
Comment (henryre) [inline multitask/multitask_tutorial.py]: nit: we've primarily been doing object type first (e.g. df_train vs. train_df). So this would be X_train_square

File: multitask/multitask_tutorial.py [hunk 22]
```diff
@@ -194,14 +210,17 @@ def split_data(data, labels, seed=123):
...
+    loss_func=partial(ce_loss, op_name="circle_head"),
+    output_func=partial(softmax, op_name="circle_head"),
     scorer=Scorer(metrics=["accuracy"]),
 )
 
 # %% [markdown]
 # Note that `Task` objects are not dependent on a particular dataset; multiple datasets can be passed through the same modules for pre-training or co-training.
 
+# %% [markdown]
+# ### Again, but faster
```
Comment (henryre) [inline multitask/multitask_tutorial.py]: Do we really need to do this big simplicity reveal? I say we just do this version, then point people to the docs or superglue for the more advanced task spec

**Comment by bhancock8** [comment]
Linked to the docs for all classes. I'm going to push back on the verbose construction section though. In all our materials, this is the place where we can go into the most detail. Defining the signature is one thing, but demonstrating it step-by-step is something that we definitely need somewhere, and I don't think it makes sense to have in the docstring. Can discuss offline and update in a subsequent PR if necessary.
